include/onnxruntime/core/framework/execution_provider.h

```diff
@@ -299,6 +300,14 @@ class IExecutionProvider {
   /** Does the EP support concurrent calls to InferenceSession::Run to execute the model.
    */
   virtual bool ConcurrentRunSupported() const { return true; }
+  /** Some session option values (default or user provided) may not work with some EPs.
+   * The EP should alter the SessionOptions object fit its implementation to avoid potential crash.
```

Comment:
This probably needs a design discussion. Not a huge fan of letting all EPs alter the session options in arbitrary ways. What if two conflict in their requirements? Whichever runs second is going to win, and one will always be broken.

I can understand EPs being able to validate whether they can run with the session options and report an issue. I also prefer returning an error instead of silently 'fixing' an invalid option as that way the user knows we're not running in the way they might think we are. 

This method also seems to be serving two purposes. One is to allow the EP to validate the options, the other is to pass them through. The naming does not call out the second usage, and the method being non-const would I think be unexpected given that. 

@pranavsharma thoughts?

Reply:
I think it would be better to let the EP copy a new one and legalize for that one only instead of the shared one. Then the behavior will not be shared between EPs.  Originally in #11138 , there is only CUDA and DML to be specially handled, so it would be a big problem at that time.

Reply:
But what if session options of ep specific has side-effect for another EP?

Reply:
@wejoncy Then it will be codesign scenario, the dev need to maintain both EPs.

Reply:
Let's take this offline in an email.